src/reversediffx.jl

```diff
@@ -0,0 +1,582 @@
+            Δ[:, start+1:start+size(xsi,2), i...]
+          end
+          start += size(xsi, 2)
+          d
+        end for xsi in xs]
+        return (Δs...,)
+    end        
+    return out_value, back
+end
+
+Base.cat(Xs::RTA...; dims) = _cat(dims, Xs...)
+Base.cat(Xs::RTV...; dims) = _cat(dims, Xs...)
+function _cat(dims, Xs::Union{RTV{<:Any, D}, RTM{<:Any, D}}...) where {D}
+    tp = tape(dims, Xs...)
+    Xs_value = value.(Xs)
+    out_value = cat(Xs_value...; dims = dims)
+    function back(Δ)
+        start = ntuple(i -> 0, Val(ndims(Δ)))
+        Δs = [begin
+          dim_xs = 1:ndims(xs)
+          till_xs = ntuple((i -> i in dims ? (i in dim_xs ? size(xs,i) : 1) : 0), Val(ndims(Δ)))
+          xs_in_Δ = ntuple(i -> till_xs[i] > 0 ? (start[i]+1:start[i]+till_xs[i]) : Colon(), Val(ndims(Δ)))
+          d = reshape(Δ[xs_in_Δ...],size(xs))
+          start = start .+ till_xs
+          d
+        end for xs in Xs]
+        return (Δs...,)
+    end        
+    out = track(out_value, D, tp)
+    record!(tp, SpecialInstruction, cat, (dims, Xs...), out, (back,))
+    return out
+end
+
+@noinline function ReverseDiff.special_reverse_exec!(instruction::SpecialInstruction{typeof(cat)})
+    output = instruction.output
+    input = instruction.input
+    input_derivs = deriv.(Base.tail(input))
+    P = instruction.cache[1]
+    jtvs = P(deriv(output))
+    for i in 1:length(jtvs)
+        input_derivs[i] .+= jtvs[i]
+    end
+    unseed!(output)
+    return nothing
+end
+
+@noinline function ReverseDiff.special_forward_exec!(instruction::SpecialInstruction{typeof(cat)})
+    output, input = instruction.output, instruction.input
+    dims = input[1]
+    Xs = value.(Base.tail(input))
+    out_value = cat(Xs..., dims = dims)
+    value!(output, out_value)
+    return nothing
+end
+
+################
+# Broadcasting #
+################
+
+"""
+    NotTracked(f::Function)
+
+A struct that can be used to wrap around closures, structs and arrays of structs declaring that they do not contain tracked variables. This enables a more efficient broadcasting of such functions and structs when doing automatic differentiation with `ReverseDiff` producing a `TrackedArray` instead of an `Array{<:TrackedReal}`.
+"""
+struct NotTracked{F} <: Function
+    f::F
+end
+(f::NotTracked{<:Union{Function, Type}})(args...; kwargs...) = f.f(args...; kwargs...)
+
+istypeorclosure(::F) where {F} = _istypeorclosure(F)
+istypeorclosure(::AbstractArray{F}) where {F} = _istypeorclosure(F)
+istypeorclosure(::Base.RefValue{F}) where {F} = _istypeorclosure(F)
+istypeorclosure(::AbstractArray{<:Real}) = false
+istypeorclosure(::Real) = false
+@generated _istypeorclosure(::Type{F}) where {F} = :($(fieldcount(F) > 0))
+
+@inline mayhavetracked(b) = istypeorclosure(b)
+@inline mayhavetracked(b::NotTracked) = false
+@inline mayhavetracked(b::Base.RefValue{<:NotTracked}) = false
+@inline mayhavetracked(b::Broadcasted) = mayhavetracked(b.f) || any(mayhavetracked, b.args)
```

Comment:
Are the `@inline` statements needed here?

Reply:
Let me see.

Reply:
Seems not.

Reply:
Done.